### PR TITLE
Arch: refactor and add unit tests for `EventStreamRuntime` docker image build

### PR DESCRIPTION
**Pull request by xingyaoww**
**What is the problem that this fixes or functionality that this introduces? Does it fix any open issues?**

**Give a brief summary of what the PR does, explaining any non-trivial design decisions**

- Refractor `EventStreamRuntime` specific logic out of `image_agnostic.py`, and create a `runtime_build.py` under `runtime/utils` to keep all the `EventStreamRuntime`-related image-agnostic docker building logic.
- Add unit tests for `runtime_build.py`.

**Other references**

### CODE REVIEW

File: opendevin/runtime/utils/runtime_build.py [hunk 1]
```diff
@@ -0,0 +1,249 @@
...
+                )
+            else:
+                logger.info(f'Building agnostic sandbox image: {target_image_name}')
+                logger.info(
+                    (
+                        f'===== Dockerfile content =====\n'
+                        f'{dockerfile_content}\n'
+                        f'==============================='
+                    )
+                )
```
Comment (Shimada666) [inline opendevin/runtime/utils/runtime_build.py:136]:
```suggestion
            if skip_init:
                logger.info(
                    f'Reusing existing od_sandbox image [{target_image_name}] but will update the source code in it.'
                )
            else:
                logger.info(f'Building agnostic sandbox image: {target_image_name}')
            logger.info(
                (
                    f'===== Dockerfile content =====\n'
                    f'{dockerfile_content}\n'
                    f'==============================='
                )
            )
```

File: opendevin/runtime/utils/runtime_build.py [hunk 2]
```diff
@@ -0,0 +1,249 @@
...
+    dockerfile_content += f'RUN mv /opendevin/{source_code_dirname} /opendevin/code\n'
+    # install (or update) the dependencies
+    dockerfile_content += (
+        'RUN cd /opendevin/code && '
+        '/opendevin/miniforge3/bin/mamba run -n base poetry env use python3.11 && '
+        '/opendevin/miniforge3/bin/mamba run -n base poetry install\n'
+        # for browser (update if needed)
+        'RUN apt-get update && cd /opendevin/code && /opendevin/miniforge3/bin/mamba run -n base poetry run playwright install --with-deps chromium\n'
+    )
+    return dockerfile_content
```
Comment (Shimada666) [inline opendevin/runtime/utils/runtime_build.py:102]: The string concatenation of Dockerfile looks a bit messy, but don't worry, we can optimize it later.

**Comment by Shimada666** [comment]
Most LGTM! Great work!

File: opendevin/runtime/utils/image_agnostic.py [hunk 3]
```diff
@@ -36,122 +38,29 @@ def generate_dockerfile(base_image: str) -> str:
...
                 file.write(dockerfile_content)
 
             api_client = docker_client.api
             build_logs = api_client.build(
-                path=temp_dir,
-                tag=target_image_name,
-                rm=True,
-                decode=True,
-                # do not use cache when skip_init is True (i.e., when we want to update the source code in the existing image)
-                nocache=skip_init,
```
Comment (yufansong) [inline opendevin/runtime/utils/image_agnostic.py:148]: I didn't remember the details. The nocache parameter are only introduced in od-runtime-client?

Comment (xingyaoww) [reply opendevin/runtime/utils/image_agnostic.py:148]: Yes, it is added in this PR: https://github.com/OpenDevin/OpenDevin/pull/2879/files#diff-63c4153a06179d7834dbe338d2de66c9b313a6bbefcf969fabdd73019f79041cR141

File: tests/unit/test_image_agnostic_util.py [hunk 4]
```diff
@@ -20,15 +20,15 @@ def test_generate_dockerfile():
 def test_get_new_image_name_legacy():
     # test non-eventstream runtime (sandbox-based)
     base_image = 'debian:11'
-    new_image_name = _get_new_image_name(base_image, is_eventstream_runtime=False)
+    new_image_name = _get_new_image_name(base_image)
     assert new_image_name == 'od_sandbox:debian__11'
 
     base_image = 'ubuntu:22.04'
```
Comment (tobitege) [inline tests/unit/test_image_agnostic_util.py:26]: Is 23.x considered as well? Might shave off some time during build?

Comment (tobitege) [reply tests/unit/test_image_agnostic_util.py:26]: Ugh.. this is test script, sorry, didn't find the right location to put my comment! 😂

File: opendevin/runtime/utils/image_agnostic.py (hunk 3 above)
Comment (yufansong) [reply opendevin/runtime/utils/image_agnostic.py:148]: Get it. Others LGTM

File: tests/unit/test_image_agnostic_util.py (hunk 4 above)
Comment (xingyaoww) [reply tests/unit/test_image_agnostic_util.py:26]: What do you mean by "23.x" - is that a particular version of the image?

Comment (tobitege) [reply tests/unit/test_image_agnostic_util.py:26]: Yes, isn't there a Ubuntu 23.x image instead of 22 maybe?

Comment (xingyaoww) [reply tests/unit/test_image_agnostic_util.py:26]: ah yes! So this unit test is more like testing the parsing / renaming part of the agnostic image. In the next PR, I'm trying to write test cases like `test_sandbox.py` that actually tests the usability of these Runtime on different images.

Comment (tobitege) [reply tests/unit/test_image_agnostic_util.py:26]: Maybe we could switch our other images from 22 to 23 in general, I put my comment in a bad place, it wasn't about the test script itself, basically. Sorry! 😂

**Pull request merged**